### DISCUSSION TITLE
CROSS_PLATFORM_DETERMINISTIC on Andorid

### DISCUSSION THREAD

**Discussion by bsrona** [jrouwe/JoltPhysics#434] [Q&A] [answered]
I run the “PerformanceTest“ on android with "CROSS_PLATFORM_DETERMINISTIC=ON", but the hash is not equal to Windows's result. Has anyone has tried with different result? Or some steps I mistake，Thanks?

here is the gradle "externaNativeBuild" build option
cmake
{
cppFlags '-std=c++17 -Wall -Werror -ffp-contract=off -ffp-model=precise -DJPH_PROFILE_ENABLED -DJPH_DEBUG_RENDERER'
arguments '-DANDROID_TOOLCHAIN=clang', '-DANDROID_STL=c++_static', '-DCROSS_PLATFORM_DETERMINISTIC=ON'
}

**Comment by jrouwe** [comment]
What is the exact settings you're compiling the windows version with? And what architecture are you running the Android build on?

Jolt is not deterministic between 32 and 64 bits builds and it has only been tested to be deterministic between the 64-bit intel version with AVX2 enabled vs the 64-bit ARM version with NEON enabled. AVX2 is not enabled by default on the Android version (see [here](https://github.com/jrouwe/JoltPhysics/blob/9b9ca187190c55a21825b90537191f4071b127bc/Build/Android/UnitTests/src/main/cpp/CMakeLists.txt#L10)) so if you're testing Android in the intel emulator then that may why it's not deterministic.

Also, which compiler versions are you using on Windows / Android? E.g. Visual Studio 2019 does not produce correct results, you need VS2022.

I did not run the PerformanceTest myself on Android as I didn't know how to package the required test files (e.g. Assets/Human.tof) in a way that would make them loadable (this is probably trivial, but I didn't spend the time to figure things out), so if you made the code to make the performance test work I'd be interested.

**Comment by bsrona** [comment]
"What is the exact settings you're compiling the windows version with?"
The cmake settings are:
cmake -S . -B VS2022_CL -G "Visual Studio 17 2022" -DCROSS_PLATFORM_DETERMINISTIC=ON -A x64 %*

"And what architecture are you running the Android build on?"
The Android Abis:
Arm64-v8a

"Also, which compiler versions are you using on Windows / Android?"
Windows:
Visual Studio 2022
Android:
NDK 25.0.8775105

"if you made the code to make the performance test work I'd be interested."
I am just interests in the cross platform determinism, So I append the hard code in PerformanceTest.cpp file to run ConvexVsMesh (walk around the problem of the files package) scene on Android.

JPH_SUPPRESS_WARNINGS_STD_BEGIN
#include "doctest.h"
JPH_SUPPRESS_WARNINGS_STD_END

TEST_CASE("PerformanceTest")
{
	const char* a = "-s=ConvexVsMesh";
	char* args[] = {NULL, const_cast<char*>(a)};
	CHECK(main(2, args) == 0);
}


I have tested PerformanceTest(ConvexVsMesh scene) with cross platform determinism on Windows, MacOS, iOS and Android.
The hash of Discrete Quality is all 0x2288ee8a2235252d on Windows, MacOS, iOS.(Fantastic Job!!!!)
The hash of Discrete Quality is always 0x67A48E9D3CF0B7B3 on Android.(I have tested on two mobile phone)

**Comment by jrouwe** [comment]
I added a PerformanceTest build for Android in #438 and when I run it it gives 0x2288ee8a2235252d as hash, so it seems to be deterministic for me (on my Pixel 6a). I'm on the same NDK version.

Can you try this new app?

**Reply by bsrona** [reply]
It worked! Great Job!
But if I run PerformanceTest with my own hard code, the hash is still 0x67A48E9D3CF0B7B3. 
Is it affect by the test framework? My hard code run the PerformanceTest in UnitTests and invoke by UnitTestFramework.cpp.

![Determinism](https://user-images.githubusercontent.com/17686412/219907976-69d17c53-260d-4f3a-9491-8b4815063c20.png)

**Reply by bsrona** [reply]
O, I See, it may be affected by other test case?

**Comment by jrouwe** [comment, marked as answer]
It took me some time to figure this one out: Both projects have a file called Layers.h with the same classes but different implementations (it contains the collision layer definitions and the filter classes). So depending on which of those 2 files the linker decides to keep the collision behavior can be completely different which obviously means the test result is different.

**Reply by bsrona** [reply]
Ok, I see. Thank you for the quick turnaround!
